Realm/RLMLogger.mm

```diff
@@ -0,0 +1,159 @@
+        case Level::info:   return RLMLogLevelInfo;
+        case Level::detail: return RLMLogLevelDetail;
+        case Level::debug:  return RLMLogLevelDebug;
+        case Level::trace:  return RLMLogLevelTrace;
+        case Level::all:    return RLMLogLevelAll;
+    }
+    REALM_UNREACHABLE();    // Unrecognized log level.
+}
+
+static NSString* levelPrefix(Level logLevel) {
+    switch (logLevel) {
+        case Level::off:
+        case Level::all:    return @"";
+        case Level::trace:  return @"Trace";
+        case Level::debug:  return @"Debug";
+        case Level::detail: return @"Detail";
+        case Level::info:   return @"Info";
+        case Level::error:  return @"Error";
+        case Level::warn:   return @"Warning";
+        case Level::fatal:  return @"Fatal";
+    }
+    REALM_UNREACHABLE();    // Unrecognized log level.
+}
+
+struct CocoaLogger : public Logger {
+    void do_log(Level level, const std::string& message) override {
+        NSLog(@"%@: %@", levelPrefix(level), RLMStringDataToNSString(message));
+    }
+};
+
+class CustomLogger : public Logger {
+public:
+    RLMLoggerFunction function;
+    void do_log(Level level, const std::string& message) override {
+        @autoreleasepool {
+            if (function) {
+                function(logLevelForLevel(level), RLMStringDataToNSString(message));
+            }
+        }
+    }
+};
+} // anonymous namespace
+
+@implementation RLMLogger {
+    std::shared_ptr<Logger> _logger;
+}
+
+typedef void(^LoggerBlock)(RLMLogLevel level, NSString *message);
+
+- (RLMLogLevel)level {
+    return logLevelForLevel(_logger->get_level_threshold());
+}
+
+- (void)setLevel:(RLMLogLevel)level {
+    _logger->set_level_threshold(levelForLogLevel(level));
+}
+
++ (void)initialize {
+    auto defaultLogger = std::make_shared<CocoaLogger>();
+    defaultLogger->set_level_threshold(Level::info);
+    Logger::set_default_logger(defaultLogger);
+}
+
+- (instancetype)initWithLogger:(std::shared_ptr<Logger>)logger {
+    if (self = [self init]) {
+        self->_logger = logger;
+    }
+    return self;
+}
+
+- (instancetype)initWithLevel:(RLMLogLevel)level logFunction:(RLMLogFunction)logFunction {
+    if (self = [super init]) {
+        auto logger = std::make_shared<CustomLogger>();
+        logger->set_level_threshold(levelForLogLevel(level));
+        logger->function = logFunction;
+        self->_logger = logger;
+    }
+    return self;
+}
+
+- (void)logWithLevel:(RLMLogLevel)logLevel message:(NSString *)message, ... {
+    auto level = levelForLogLevel(logLevel);
+    if (_logger->would_log(level)) {
+        va_list args;
+        va_start(args, message);
+        _logger->log(level, message.UTF8String, args);
+        va_end(args);
+    }
+}
+
+- (void)logLevel:(RLMLogLevel)logLevel message:(NSString *)message {
+    auto level = levelForLogLevel(logLevel);
+    if (_logger->would_log(level)) {
+        _logger->log(level, message.UTF8String);
```

Comment:
The second argument to `Logger:log()` is a format string, so this'll break if `message` contains the character `%`. It needs to instead be `_logger->log(level, "%1", message.UTF8String);` so that it doesn't parse `message`.